### PR TITLE
Python 3 fixes

### LINKED ISSUE #55: Python 3 issues
[flake8](http://flake8.pycqa.org) testing of https://github.com/norvig/pytudes on Python 3.7.0

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./py/testaccum.py:21:21: E999 SyntaxError: invalid syntax
    print acc_display
                    ^
./py/py2html.py:79:12: F821 undefined name 'cmp'
    return cmp(num(x[0]), num(y[0]))
           ^
./py/lispy.py:119:26: E999 SyntaxError: invalid syntax
            print '%s: %s' % (type(e).__name__, e)
                         ^
./py/pal.py:130:22: E999 SyntaxError: invalid syntax
            print self.best
                     ^
./py/lispytest.py:106:19: E999 SyntaxError: invalid syntax
            print x, '=>', to_string(result)
                  ^
./py/ibol.py:78:47: E999 SyntaxError: invalid syntax
    print "Number of genomes: %d (%d distinct)" % (len(genomes), len(set(genomes)))
                                              ^
./py/docex.py:101:16: E999 TabError: inconsistent use of tabs and spaces in indentation
	if self.failed:
               ^
./py/lettercount.py:64:49: E999 SyntaxError: invalid syntax
        print '%s: %s %s words (%s tokens) at %s' % (
                                                ^
./py/parse.py:29:19: E999 TabError: inconsistent use of tabs and spaces in indentation
		rhs = mklist(rhs)
                  ^
./py/ngrams.py:30:28: E999 SyntaxError: invalid syntax
    print 'Running tests...'
                           ^
./py/pal2.py:188:37: E999 SyntaxError: invalid syntax
        print 'using reversibles ...'
                                    ^
./py/lis.py:98:26: F821 undefined name 'raw_input'
        val = eval(parse(raw_input(prompt)))
                         ^
./py/yaptu.py:76:49: E999 SyntaxError: invalid syntax
            if self.verbose: print '=== eval{%s}' % expr,
                                                ^
11    E999 TabError: inconsistent use of tabs and spaces in indentation
2     F821 undefined name 'raw_input'
13
```

### PR DESCRIPTION
**Pull request by cclauss**
Partially fixes #55  Fix many but not all Python 3 incompatibilities.

Four issues remain.  The first two are related to lambda syntax in Python 3 only.  The second two fail in exactly the same way when the [flake8](http://flake8.pycqa.org) tests are run on Python 2.

$ __flake8 . --count --show-source --statistics --select=E901,E999,F821,F822,F823__
```
./py/ngrams.py:216:42: E999 SyntaxError: invalid syntax
    c, edit = max(candidates, key=lambda (c,e): Pedit(e) * Pw(c))
                                         ^
./py/lispy.py:168:30: F821 undefined name 'file'
     'port?': lambda x:isa(x,file), 'apply':lambda proc,l: proc(*l),
                             ^
./py/lettercount.py:96:44: F821 undefined name 'keywords'
def tag(name, **kwds): return '<' + name + keywords(kwds) + '>'
                                           ^
./py/lettercount.py:152:57: F821 undefined name 'all_lengths'
        return sum(getcount(counts, s, pos, L) for L in all_lengths)
                                                        ^
1     E999 SyntaxError: invalid syntax
3     F821 undefined name 'keywords'
4
```
The [flake8](http://flake8.pycqa.org) tests try to find syntax errors and undefined names.

__E901,E999,F821,F822,F823__ are the "_showstopper_" flake8 issues that can halt the runtime with a SyntaxError, NameError, etc. Most other flake8 issues are merely "style violations" -- useful for readability but they do not effect runtime safety.
* F821: undefined name `name`
* F822: undefined name `name` in `__all__`
* F823: local variable name referenced before assignment
* E901: SyntaxError or IndentationError
* E999: SyntaxError -- failed to compile a file into an Abstract Syntax Tree

### CODE REVIEW

File: py/pal2.py [hunk 1]
```diff
@@ -202,8 +208,8 @@ def report(self):
         if N > self.best and (N > 12500 or N > self.best+500):
             self.best = len(self)
             self.bestphrase = str(self)
-            print '%5d phrases (%5d words) in %3d seconds' % (
-                self.best, self.bestphrase.count(' ')+1, time.clock() - self.starttime)
+            print('%5d phrases (%5d words) in %3d seconds' % (
+                self.best, self.bestphrase.count(' ')+1, time.clock() - self.starttime))
             assert is_panama(self.bestphrase)
             f = open('pallog%d.txt' % (id(self) % 10000), 'w')
             f.write(self.bestphrase + '\n')
```
Comment (jona-sassenhagen) [inline py/pal2.py]:
```suggestion
            with open('pallog%d.txt' % (id(self) % 10000), 'w') as f:
                f.write(self.bestphrase + '\n')
```

File: py/yaptu.py [hunk 2]
```diff
@@ -155,10 +156,10 @@ def copyfile(self, filename, ext="html"):
         "Convert filename.* to filename.ext, where ext defaults to html."
         global yaptu_filename
         outname = re.sub('[.][a-zA-Z0-9]+?$', '', filename) + '.'+ext
-        print 'Transforming', filename, 'to', outname
-        self.globals['_bl'] = file(filename).readlines()
+        print('Transforming', filename, 'to', outname)
+        self.globals['_bl'] = open(filename).readlines()
         yaptu_filename = filename
-        self.outf = file(outname, 'w')
+        self.outf = open(outname, 'w')
```
Comment (jona-sassenhagen) [inline py/yaptu.py]: maybe also use `with open(...` here, too ...

File: py/pal2.py (hunk 1 above)
Comment (cclauss) [reply py/pal2.py]: I always use the __with open() as__ syntax in my own code but this is not my own code.

Given the genius of this code's original author, my preference is to make only those minimal changes required to achieve Python 3 compatibility.

File: py/yaptu.py (hunk 2 above)
Comment (cclauss) [reply py/yaptu.py]: Same.

File: py/pal2.py (hunk 1 above)
Comment (jona-sassenhagen) [reply py/pal2.py]: Understandable. So keep unless @norvig thinks it should be updated.

**Pull request merged**